samza-core/src/main/java/org/apache/samza/container/grouper/stream/AbstractSystemStreamPartitionGrouper.java

```diff
@@ -0,0 +1,186 @@
+
+  private final StreamPartitionMapper streamPartitionMapper;
+  private final Set<SystemStreamPartition> broadcastSystemStreamPartitions;
+
+  public AbstractSystemStreamPartitionGrouper(Config config) {
+    Preconditions.checkNotNull(config);
+    this.broadcastSystemStreamPartitions = new TaskConfigJava(config).getBroadcastSystemStreamPartitions();
+    this.streamPartitionMapper = getPartitionExpansionAlgorithm(config);
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * 1. Invokes the sub-class {@link org.apache.samza.container.grouper.task.TaskNameGrouper#group(Set)} method to generate the task to partition assignments.
+   * 2. Uses the {@link StreamPartitionMapper} to compute the previous {@link SystemStreamPartition} for every partition of the expanded streams.
+   * 3. Uses the previous, current task to partition assignments and result of {@link StreamPartitionMapper} to redistribute the expanded {@link SystemStreamPartition}'s
+   * to correct tasks after the stream expansion.
+   */
+  @Override
+  public Map<TaskName, Set<SystemStreamPartition>> group(Set<SystemStreamPartition> ssps, GrouperContext grouperContext) {
+    Map<TaskName, Set<SystemStreamPartition>> currentTaskAssignments = group(ssps);
+
+    if (grouperContext.getPreviousTaskToSSPAssignment().isEmpty()) {
+      LOGGER.info("Previous task to partition assignment does not exist. Using the result from the group method.");
+      return currentTaskAssignments;
+    }
+
+    Map<SystemStreamPartition, TaskName> previousSSPToTask = getPreviousSSPToTaskMapping(grouperContext);
+
+    Map<TaskName, PartitionGroup> taskToPartitionGroup = new HashMap<>();
+    currentTaskAssignments.forEach((taskName, systemStreamPartitions) -> taskToPartitionGroup.put(taskName, new PartitionGroup(taskName, systemStreamPartitions)));
+
+    Map<SystemStream, Integer> previousStreamToPartitionCount = getStreamToPartitionCount(grouperContext.getPreviousTaskToSSPAssignment());
+    Map<SystemStream, Integer> currentStreamToPartitionCount = getStreamToPartitionCount(currentTaskAssignments);
+
+    for (Map.Entry<TaskName, Set<SystemStreamPartition>> entry : currentTaskAssignments.entrySet()) {
+      TaskName currentlyAssignedTask = entry.getKey();
+      for (SystemStreamPartition currentSystemStreamPartition : entry.getValue()) {
+        if (broadcastSystemStreamPartitions.contains(currentSystemStreamPartition)) {
+          LOGGER.info("SystemStreamPartition: {} is part of broadcast stream. Skipping reassignment.", currentSystemStreamPartition);
+        } else {
+          SystemStream systemStream = currentSystemStreamPartition.getSystemStream();
+
+          Integer previousStreamPartitionCount = previousStreamToPartitionCount.getOrDefault(systemStream, 0);
+          Integer currentStreamPartitionCount = currentStreamToPartitionCount.getOrDefault(systemStream, 0);
+
+          if (previousStreamPartitionCount > 0 && previousStreamPartitionCount < currentStreamPartitionCount
```

Comment:
Could we also make this work both incase of a partition-expansion and reduction.